content/5_Plat/HLD.mdx

```diff
@@ -15,6 +15,218 @@ frequency: 1
 
 ## Tutorial
 
+In some advanced problems on trees, the problem may ask you to support the following operations:
+
+- Update all nodes along the path from node $x$ to node $y$. (Note that this path is unique due to properties of trees).
+
+- Find the sum, maximum, minimum (or any other operation that satisfies the associative property) along the path from node $x$ to node $y$.
+
+Other simpler algorithms such as difference on trees or LCAs could be used to support any one of the operations. However, Heavy Light Decomposition (or HLD) is needed to support both. 
+
+**Definitions:**
+
+To understand HLD, I will first clarify the definition of these important terms: 
+
+- A  **heavy child** of a node is the child with the largest subtree size rooted at the child.
+- A  **light child** of a node is any child that is not a **heavy child**.
+- A  **heavy edge** connects the current node to its **heavy child**.
+- A  **light edge** connects the current node to any of its **light children**.
+- A  **heavy path** is the path formed by a collection **heavy edges**.
+- A  **light path** is the path formed by a collection **light edges**.
+
+**Properties:**
+
+Given the definitions stated above, an important property can be derived (I will prove this a bit later).
+
+- Any path from node $x$ to node $y$ on the tree can pass through at most $\mathcal{O}(\log N)$ light edges.
+
+Lets suppose the above property is true for a moment. Since a heavy path can only be broken by a light edge (or else the edge will be a part of the heavy path), we can know that there are at most $\mathcal{O}(\log N)$ heavy paths on any path from an arbitrary node $x$ to an arbitrary node $y$.
+
+In addition, by using segment trees (or any other RURQ data structure) we can calculate the value of any consecutive interval on any heavy path in $\mathcal{O}(\log N)$ time. 
+
+Since there are at most $\mathcal{O}(\log N)$ heavy paths and $\mathcal{O}(\log N)$ light edges, computing the value on the path from node $x$ to node $y$ will take $\mathcal{O}(\log^2 N + \log N)$ = $\mathcal{O}(\log^2 N)$ time. We can answer $Q$ queries in $\mathcal{O}(Q \log^2 N)$ time if the above property is true. This computational complexity allows us to pass questions where $N$ = $Q$ = $10^5$.
```

Comment:
> This computational complexity allows us to pass questions where $N$ = $Q$ = $10^5$.

i'm not so sure about this

Reply:
Ok I can remove that statement. But just for future reference O(N * log^2 N) roughly equals 2.7 * 10^7 when N = 10^5 so this complexity is good right ?

Reply:
yea